modules/standard/Allocators.chpl

```diff
@@ -0,0 +1,377 @@
+
+  Custom allocators can be defined as classes or records that implement the
+  ``allocator`` interface. The interface has two methods: ``allocate`` and
+  ``deallocate``. The ``allocate`` method accepts an integer ``n`` and returns
+  a :type:`~CTypes.c_ptr` to the allocated memory. The ``deallocate``
+  method accepts a :type:`~CTypes.c_ptr` to the allocated memory. Allocators are
+  free to implement their own memory management strategies.
+
+  Limitations:
+    * The ``newWithAllocator`` and ``deleteWithAllocator`` procedures are meant
+      to be stand-ins pending a more elegant syntax to use custom allocators.
+    * This module currently only supports allocating Chapel classes. In the
+      future we hope to support other heap objects like arrays.
+    * Allocating managed (:type:`~OwnedObject.owned`/:type:`~SharedObject.shared`)
+      Chapel classes is not supported, only ``unmanaged`` classes are supported.
+
+*/
+@unstable(category="experimental", reason="The Allocators module is under development and does not have a stable interface yet")
+module Allocators {
+  private use CTypes;
+  private use Reflection;
+  import ChapelLocks;
+  private use ChplConfig;
+
+  private inline proc alignup(ptr: c_ptr(void), param alignment): c_ptr(void)
+    where alignment > 0 && (alignment & (alignment - 1)) == 0 {
+
+    param mask = alignment - 1;
+    const iptr = ptr:c_intptr;
+    const alignedPtr = (iptr + mask) & ~mask;
+    return alignedPtr:c_ptr(void);
+  }
+
+  // TODO: interfaces are not supported by chpldoc yet: https://github.com/chapel-lang/chapel/issues/17383
+  /*
+    All allocators must implement this interface. The interface has two
+    methods: ``allocate`` and ``deallocate``. The ``allocate`` method accepts
+    an integer size and returns a pointer to the allocated memory. The
+    ``deallocate`` method accepts a pointer to the allocated memory.
+
+    Allocators may either be classes or records.
+  */
+  interface allocator {
+    /*
+      Allocate memory for ``n`` bytes.
+
+      :arg n: The number of bytes to allocate.
+      :returns: A ``c_ptr(void)`` to the allocated memory.
+    */
+    proc ref Self.allocate(n: int): c_ptr(void);
+    /*
+      Deallocate memory at ``p``.
+
+      :arg p: The pointer to the allocated memory.
+    */
+    proc ref Self.deallocate(p: c_ptr(void));
+  }
+
+  private proc newTypeCheckHelper(type T) param {
+    if !isClassType(T) then
+      compilerError(T:string + " is not a class", 2);
+    if !isUnmanagedClassType(T) then
+      compilerError(T:string + " is not unmanaged", 2);
+  }
+  private proc checkInterfaceHelper(alloc) param {
+    if __primitive("implements interface", alloc, allocator) == 2 {
+      compilerError(alloc.type:string + " does not implement 'allocator'", 2);
+    }
+  }
+
+  /*
+    Allocate a new unmanaged class with type ``T`` by invoking the ``allocate``
+    method of the given ``alloc``. This is a drop-in replacement for ``new``.
```

Comment:
Please add a comment that notes that named argument passing will not work with this function at present.